src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -244,14 +222,11 @@ private void FreeHandle(IntPtr handle)
             {
                 if (_handleToContextMap.TryRemove(handle, out _))
                 {
-                    _outstandingHandles = IntPtr.Subtract(_outstandingHandles, 1);
-                    Debug.Assert(_outstandingHandles.ToInt64() >= 0, $"Unexpected _outstandingHandles: {_outstandingHandles}");
-
                     //
                     // If we've allocated all possible handles for this instance, and freed them all, then
                     // we don't need the event loop any more, and can reclaim resources.
                     //
-                    if (IsFull && _outstandingHandles == IntPtr.Zero)
+                    if (IsFull && _handleToContextMap.IsEmpty)
```

Comment:
ConcurrentDictionary.IsEmpty isn't the cheapest of operations.  But presumably it'll be called here very rarely, since it's only called when an engine is full, which is after a few billions sockets, right? :)

Reply:
Yes :)